components/threading/virtualized_thread_viewer/virtualized_thread_viewer.tsx

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {PureComponent, RefObject} from 'react';
+import AutoSizer from 'react-virtualized-auto-sizer';
+import {DynamicSizeList} from 'dynamic-virtualized-list';
+import memoize from 'memoize-one';
+
+import {$ID} from 'mattermost-redux/types/utilities';
+import {Channel} from 'mattermost-redux/types/channels';
+import {Post} from 'mattermost-redux/types/posts';
+import {UserProfile} from 'mattermost-redux/types/users';
+
+import {Posts} from 'mattermost-redux/constants';
+import {isDateLine, isStartOfNewMessages} from 'mattermost-redux/utils/post_list';
+
+import DelayedAction from 'utils/delayed_action';
+import * as Utils from 'utils/utils.jsx';
+import Constants from 'utils/constants';
+import {FakePost} from 'types/store/rhs';
+import {getNewMessageIndex, getPreviousPostId, getLatestPostId} from 'utils/post_utils';
+
+import FloatingTimestamp from 'components/post_view/floating_timestamp';
+import {THREADING_TIME as BASE_THREADING_TIME} from 'components/threading/common/options';
+
+import CreateComment from './create_comment';
+import Row from './thread_viewer_row';
+
+export type OwnProps = {
+    channel: Channel;
+    onCardClick: (post: Post) => void;
+    onCardClickPost: (post: Post) => void;
+    openTime: number;
+    postIds: Array<$ID<Post | FakePost>>;
+    removePost: (post: Post) => void;
+    selected: Post | FakePost;
+    useRelativeTimestamp: boolean;
+};
+
+type Props = {
+    currentUserId: string;
+    directTeammate: UserProfile | undefined;
+    highlightedPostId?: $ID<Post>;
+    lastPost: Post;
+    previewCollapsed: string;
+    previewEnabled: boolean;
+    replyListIds: string[];
+    teamId: string;
+} & OwnProps;
+
+type State = {
+    isMobile: boolean;
+    isScrolling: boolean;
+    topRhsPostId?: string;
+    userScrolled: boolean;
+    userScrolledToBottom: boolean;
+}
+
+const virtListStyles = {
+    position: 'absolute',
+    top: '0',
+    maxHeight: '100%',
+    willChange: 'auto',
+};
+
+const THREADING_TIME: typeof BASE_THREADING_TIME = {
+    ...BASE_THREADING_TIME,
+    units: [
+        'now',
+        'minute',
+        'hour',
+        'day',
+        'week',
+        'month',
+        'year',
+    ],
+};
+
+const OFFSET_TO_SHOW_TOAST = Number(-50);
+const OVERSCAN_COUNT_FORWARD = 30;
+const OVERSCAN_COUNT_BACKWARD = 30;
+
+class ThreadViewerVirtualized extends PureComponent<Props, State> {
+    private mounted = false;
+    private scrollStopAction: DelayedAction;
+    postCreateContainerRef: React.RefObject<HTMLDivElement>;
+    listRef: RefObject<DynamicSizeList>;
+    innerRef: RefObject<HTMLDivElement>;
+    initRangeToRender: number[];
+
+    constructor(props: Props) {
+        super(props);
+
+        this.scrollStopAction = new DelayedAction(this.handleScrollStop);
+        this.listRef = React.createRef();
```

Comment:
Is the repetition on line 106 and 107 intended?

Reply:
Oh.. not at all :blush: thanks!